### PR TITLE
Feature/gh 10108 update config cypress test

### LINKED ISSUE #10108: Cypress test
See update test for VC and implement in a similar way

### PR DESCRIPTION
**Pull request by AlexanderKuschner**
Cypress test for Update Configuration;
Possibility to de-selectt all values

### CODE REVIEW

**Comment by Uli-Tiger** [comment]
@Larisa-Staroverova : can you review the cypress part?

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.spec.ts [hunk 1]
```diff
@@ -7,13 +7,15 @@ import { Configurator } from './../core/model/configurator.model';
...
-const attributeName = 'AttrName';
+const attributeName = '9999';
 const selectedSingleValue = 'SomeValue';
 const value1: Configurator.Value = { valueCode: 'ValueCode1', selected: true };
 const value2: Configurator.Value = { valueCode: 'ValueCode2', selected: false };
 const value3: Configurator.Value = { valueCode: 'ValueCode3', selected: true };
 const userInput = '123';
 const expectedValueIdsMulti = 'ValueCode1,ValueCode3,';
+const expectedValueIdsMultiNoSelection = ',';
+const expectedProcessedSingleValueNoValue = ',';
```
Comment (Uli-Tiger) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.spec.ts:18]: expecting just a comma, when no value is selected seems odd to me?

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts [hunk 2]
```diff
@@ -62,14 +63,28 @@ export class CpqConfiguratorSerializer
...
-    });
+
+    if (selectedValues && selectedValues.length > 0) {
+      valueIds = '';
+      selectedValues.forEach((value) => {
+        valueIds += value.valueCode + VALUE_SEPARATOR;
+      });
+    } else {
+      valueIds = VALUE_SEPARATOR;
+    }
```
Comment (Uli-Tiger) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts]: The else part is strange. just returning the Value Separator itself, when there is no value seems odd to me?
Would have expected the empty string, or just some special value.

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]
@AlexanderKuschner : Can you check my remarks?

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.spec.ts (hunk 1 above)
Comment (Larisa-Staroverova) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.spec.ts:18]: Furthermore why did you define 2 variables for the same value?

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts [hunk 3]
```diff
@@ -62,14 +63,28 @@ export class CpqConfiguratorSerializer
     return updateAttribute;
   }
 
+  processSelectedSingleValue(singleValue: string): string {
+    let processedValue: string = singleValue;
+    if (!processedValue) {
+      processedValue = VALUE_SEPARATOR;
```
Comment (Larisa-Staroverova) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts:70]: Why should the value separator be added when no value is processed?

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts [hunk 4]
```diff
@@ -42,8 +42,9 @@ export class CpqConfiguratorSerializer
       attribute.uiType === Configurator.UiType.RADIOBUTTON_PRODUCT ||
       attribute.uiType === Configurator.UiType.SINGLE_SELECTION_IMAGE
     ) {
-      updateAttribute.changeAttributeValue.attributeValueIds =
-        attribute.selectedSingleValue;
+      updateAttribute.changeAttributeValue.attributeValueIds = this.processSelectedSingleValue(
```
Comment (Larisa-Staroverova) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts:45]: do you mean _change**d**AttributeValue_?

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts [hunk 5]
```diff
@@ -62,14 +63,28 @@ export class CpqConfiguratorSerializer
...
+    let valueIds;
     const selectedValues: Configurator.Value[] = attribute.values.filter(
       (value) => value.selected
     );
-    selectedValues.forEach((value) => {
-      valueIds += value.valueCode + VALUE_SEPARATOR;
-    });
+
+    if (selectedValues && selectedValues.length > 0) {
+      valueIds = '';
```
Comment (Larisa-Staroverova) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts]: You don't need to set this variable to empty

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts [hunk 6]
```diff
@@ -62,14 +63,28 @@ export class CpqConfiguratorSerializer
...
+    let processedValue: string = singleValue;
+    if (!processedValue) {
+      processedValue = VALUE_SEPARATOR;
+    }
+    return processedValue;
+  }
+
   prepareValueIds(attribute: Configurator.Attribute): string {
-    let valueIds = '';
+    let valueIds;
```
Comment (Larisa-Staroverova) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts]: If you don't set the value to the variable, you have to define the type of it. In this case: _**let** valueIds: **string**;_

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts (hunk 2 above)
Comment (Larisa-Staroverova) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts]: I do have the same question why it is not possible to send an empty string to CPQ

File: projects/storefrontapp-e2e-cypress/cypress/helpers/product-configuration.ts [hunk 7]
```diff
@@ -1043,3 +1043,33 @@ export function login(email: string, password: string, name: string): void {
...
   cy.get('.cx-login-greet').should('contain', name);
 }
+
+/**
+ * Verifies whether the radiobutton is selected.
+ *
+ * @param {string} attributeName - Attribute name
+ * @param {string} valueName - Value name
+ */
+export function verifyRadioButtonSelected(
```
Comment (Larisa-Staroverova) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/product-configuration.ts:1053]: Due to the fact that there are some methods that do almost the same, for instance: **isCheckboxSelected**, **isImageSelected**, **isImageNotSelected**. Could you rename your methods from **verifyRadioButtonSelected** to **isRadioButtonSelected** & from **verifyCheckboxNotSelected** to **isCheckboxNotSelected**?

Comment (Larisa-Staroverova) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/product-configuration.ts:1053]: Then a task should be defined on VariantConfiguration board to refactor these methods to only 2 general methods, namely **isSelected** & **isNotSelected** independently from the UI type & downported to cpq epic

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/cpq-configuration.e2e-spec.ts [hunk 8]
```diff
@@ -38,4 +53,67 @@ context('CPQ Configuration', () => {
...
+      configuration.isAttributeDisplayed(
+        ATTR_COFFEE_MACHINE_CUPS_DAY,
+        radioGroup
+      );
+      configuration.selectAttribute(
+        ATTR_COFFEE_MACHINE_CUPS_DAY,
+        radioGroup,
+        COFFEE_MACHINE_CUPS_DAY_300_500
+      );
+      configuration.verifyRadioButtonSelected(
```
Comment (Larisa-Staroverova) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/cpq-configuration.e2e-spec.ts:70]: Please, rename to **isRadioButtonSelected**

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/cpq-configuration.e2e-spec.ts [hunk 9]
```diff
@@ -38,4 +53,67 @@ context('CPQ Configuration', () => {
...
+
+    it('should support update values for checkbox list attribute type', () => {
+      goToPDPage(testProductCoffeeMachine);
+      configuration.clickOnConfigureBtnInCatalog();
+      configuration.isAttributeDisplayed(
+        ATTR_COFFEE_MACHINE_STARB_MODE,
+        checkBoxList
+      );
+
+      configuration.verifyCheckboxNotSelected(
```
Comment (Larisa-Staroverova) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/cpq-configuration.e2e-spec.ts:93]: Please, rename to **isCheckboxNotSelected**

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts (hunk 2 above)
Comment (AlexanderKuschner) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts]: This is exactly the value which is expected from the CPQ update API to remove all / last values. Other values (e.g. empty string) leads to the "bad request".

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.spec.ts (hunk 1 above)
Comment (AlexanderKuschner) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.spec.ts:18]: This is exactly the value which is expected from the CPQ update API to remove all / last values. Other values (e.g. empty string) leads to the "bad request".

**Review by Larisa-Staroverova — CHANGES_REQUESTED** [review]

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts (hunk 3 above)
Comment (AlexanderKuschner) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts:70]: This is exactly the value which is expected from the CPQ update API to remove all / last values. Other values (e.g. empty string) leads to the "bad request".

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts (hunk 4 above)
Comment (AlexanderKuschner) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-serializer.ts:45]: this is the documented parameter name in the API

**Review by Larisa-Staroverova — APPROVED** [review]

**Review by Uli-Tiger — APPROVED** [review]

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]
can you plz put a comment into the code that the separator is required to remove the value

**Comment by AlexanderKuschner** [comment]
> can you plz put a comment into the code that the separator is required to remove the value

Done

**Pull request merged**